engine/src/main/java/org/pentaho/di/trans/step/BaseStep.java

```diff
@@ -1839,106 +1840,109 @@ protected void waitUntilTransformationIsStarted() {
+        try {
+          Thread.sleep( 0, 1 );
+        } catch ( InterruptedException e ) {
+          // Ignore sleep interruption exception
+        }
       }
-    }
 
-    // See if this step is receiving partitioned data...
-    // In that case it might be the case that one input row set is receiving
-    // all data and
-    // the other rowsets nothing. (repartitioning on the same key would do
-    // that)
-    //
-    // We never guaranteed that the input rows would be read one by one
-    // alternatively.
-    // So in THIS particular case it is safe to just read 100 rows from one
-    // rowset, then switch to another etc.
-    // We can use timeouts to switch from one to another...
-    //
-    while ( row == null && !isStopped() ) {
-      // Get a row from the input in row set ...
-      // Timeout immediately if nothing is there to read.
-      // We will then switch to the next row set to read from...
+      // See if this step is receiving partitioned data...
+      // In that case it might be the case that one input row set is receiving
+      // all data and
+      // the other rowsets nothing. (repartitioning on the same key would do
+      // that)
       //
-      row = inputRowSet.getRowWait( 1, TimeUnit.MILLISECONDS );
-      if ( row != null ) {
-        incrementLinesRead();
-        blockPointer++;
-      } else {
-        // Try once more...
-        // If row is still empty and the row set is done, we remove the row
-        // set from
-        // the input stream and move on to the next one...
+      // We never guaranteed that the input rows would be read one by one
+      // alternatively.
+      // So in THIS particular case it is safe to just read 100 rows from one
+      // rowset, then switch to another etc.
+      // We can use timeouts to switch from one to another...
+      //
+      while ( row == null && !isStopped() ) {
+        // Get a row from the input in row set ...
+        // Timeout immediately if nothing is there to read.
+        // We will then switch to the next row set to read from...
         //
-        if ( inputRowSet.isDone() ) {
-          row = inputRowSet.getRowWait( 1, TimeUnit.MILLISECONDS );
-          if ( row == null ) {
-            inputRowSets.remove( currentInputRowSetNr );
-            if ( inputRowSets.isEmpty() ) {
-              return null; // We're completely done.
+        row = inputRowSet.getRowWait( 1, TimeUnit.MILLISECONDS );
+        if ( row != null ) {
+          incrementLinesRead();
+          blockPointer++;
+        } else {
+          // Try once more...
+          // If row is still empty and the row set is done, we remove the row
+          // set from
+          // the input stream and move on to the next one...
+          //
+          if ( inputRowSet.isDone() ) {
+            row = inputRowSet.getRowWait( 1, TimeUnit.MILLISECONDS );
+            if ( row == null ) {
+              inputRowSets.remove( currentInputRowSetNr );
```

Comment:
This is problematic - you are mutating the inputRowSets without using the synchronization block. As said above, think this would be a write lock here...

Reply:
This is actually inside a synchronized block. See https://github.com/pentaho/pentaho-kettle/pull/5118/files/1199c579aef9506c52ed8c1b5b07cafae7117b30#diff-9a94b8b1359d88cee88bab433700873dR1846